src/core/dbus-util.c

```diff
@@ -228,3 +228,26 @@ int bus_read_mount_options(
 
         return 0;
 }
+
+int bus_property_get_activation_details(sd_bus_message *reply, ActivationDetails *details) {
+        _cleanup_strv_free_ char **pairs = NULL;
+        int r;
+
+        assert(reply);
+
+        r = activation_details_append_pair(details, &pairs);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_open_container(reply, 'a', "(ss)");
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH_PAIR(key, value, pairs) {
+                r = sd_bus_message_append(reply, "(ss)", *key, *value);
+                if (r < 0)
+                        return r;
+        }
+
+        return sd_bus_message_close_container(reply);
+}
```

Comment:
one of these days we should add sd_bus_message_append_strv_pairs(). I have the suspicion this is not the only place where we do this.

Reply:
that would be good, but this is already large-ish PR and I'd rather not add more functionality